protocol/chainlib/provider_node_subscription_manager.go

```diff
@@ -0,0 +1,520 @@
+	msg, ok := data.(*rpcclient.JsonrpcMessage)
+	if !ok {
+		return nil, fmt.Errorf("data is not a *rpcclient.JsonrpcMessage, but: %T", data)
+	}
+
+	var convertedMsg any
+	var err error
+
+	switch apiCollection.GetCollectionData().ApiInterface {
+	case spectypes.APIInterfaceJsonRPC:
+		convertedMsg, err = rpcInterfaceMessages.ConvertJsonRPCMsg(msg)
+		if err != nil {
+			return nil, err
+		}
+	case spectypes.APIInterfaceTendermintRPC:
+		convertedMsg, err = rpcInterfaceMessages.ConvertTendermintMsg(msg)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("unsupported API interface: %s", apiCollection.GetCollectionData().ApiInterface)
+	}
+
+	marshalledMsg, err := gojson.Marshal(convertedMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	return marshalledMsg, nil
+}
+
+func (pnsm *ProviderNodeSubscriptionManager) signReply(ctx context.Context, reply *pairingtypes.RelayReply, consumerAddr sdk.AccAddress, chainMessage ChainMessage, request *pairingtypes.RelayRequest) error {
+	// Send the first setup message to the consumer in a go routine because the blocking listening for this channel happens after this function
+	dataReliabilityEnabled, _ := pnsm.chainParser.DataReliabilityParams()
+	blockLagForQosSync, averageBlockTime, blockDistanceToFinalization, blocksInFinalizationData := pnsm.chainParser.ChainBlockStats()
+	relayTimeout := GetRelayTimeout(chainMessage, averageBlockTime)
+
+	if dataReliabilityEnabled {
+		var err error
+		latestBlock, _, requestedHashes, modifiedReqBlock, _, updatedChainMessage, err := pnsm.relayFinalizationBlocksHandler.GetParametersForRelayDataReliability(ctx, request, chainMessage, relayTimeout, blockLagForQosSync, averageBlockTime, blockDistanceToFinalization, blocksInFinalizationData)
+		if err != nil {
+			return err
+		}
+
+		err = pnsm.relayFinalizationBlocksHandler.BuildRelayFinalizedBlockHashes(ctx, request, reply, latestBlock, requestedHashes, updatedChainMessage, relayTimeout, averageBlockTime, blockDistanceToFinalization, blocksInFinalizationData, modifiedReqBlock)
+		if err != nil {
+			return err
+		}
+	}
+
+	var ignoredMetadata []pairingtypes.Metadata
+	reply.Metadata, _, ignoredMetadata = pnsm.chainParser.HandleHeaders(reply.Metadata, chainMessage.GetApiCollection(), spectypes.Header_pass_reply)
+	reply, err := lavaprotocol.SignRelayResponse(consumerAddr, *request, pnsm.privKey, reply, dataReliabilityEnabled)
+	if err != nil {
+		return err
+	}
+	reply.Metadata = append(reply.Metadata, ignoredMetadata...) // appended here only after signing
+	return nil
+}
+
+func (pnsm *ProviderNodeSubscriptionManager) handleNewNodeMessage(ctx context.Context, hashedParams string, nodeMsg interface{}) {
+	pnsm.lock.RLock()
+	defer pnsm.lock.RUnlock()
+
+	if _, ok := pnsm.activeSubscriptions[hashedParams]; !ok {
+		utils.LavaFormatDebug("No hashed params in handleNewNodeMessage, connection might have been closed", utils.LogAttr("hash", hashedParams))
```

Comment:
terminate then? otherwise it keeps doing this every node message without sending to consumers

Reply:
this is ok, it will terminate on its own next case, this is similar to the race case for the ticker we had. this solution is valid.